working/2509-private-named-parameters/feature-specification.md

```diff
@@ -431,25 +464,85 @@ main() {
 }
 ```
 
-*Note that the proposal only applies named parameters and only to ones which are
+*Note that this section only applies to named parameters that are also
 initializing formals or field parameters. A named parameter can only have a
 private name in a context where it is _useful_ to do so because it corresponds
 to a private instance field. For all other named parameters it is still a
 compile-time error to have a private name.*
 
+### Private positional parameters
+
+Dart already allows a positional parameter to have a private name. This is
+useful for a positional constructor parameter that is an initializing formal or
+a declaring parameter for a private instance variable. For other parameters,
+it's non-idiomatic but harmless to give them private names. *(Dart also allows
+local variables to have private names, which is unusual but harmless. A lint
+suggests that users don't do this.)*
+
+However, the user experience of the language is more than just the semantics.
+Generated documentation, code navigation in doc comment references, and runtime
+error messages all reflect a function or constructor's parameters back to the
+user.
+
+In all of those contexts, we want it to be an implementation detail that a
+positional parameter happens to have a private name. To that end:
+
+Given a positional parameter with private name *p* in formal parameter list L:
+
+*   If *p* has corresponding public name *n* and no other parameter in L has
+    declared name *n*, then the access name of the parameter is *n*.
+
+Since a positional parameter can't be passed using a named argument, this
+doesn't affect the language semantics. It does mean that generated docs, error
+messages, inferred super parameter names, etc. should all use the public name
+of the parameter. For example:*
+
+```dart
+const int foo = 0;
+
+class C {
+  int _foo;
+
+  /// Uses [foo] to set [_foo].
+  C(this._foo);
+}
+```
+
+Here, the generated documentation for C should show the constructor signature
+like `C(int foo)`. If a user clicks the `[foo]` reference in the doc comment
+and navigates to its definition, their IDE should take them to the constructor
+parameter `_foo` and not the unrelated top level constant `foo`.
+
+In short, a class author should never feel that can't use a private name for
+a positional initializing formal or declaring parameter because doing so might
+degrade the user experience of anyone working with the class or reveal the
+implementation detail that the parameter happens to initialize a private field.
+
+*Unlike with named parameters, this section applies to all positional
+parameters, regardless of whether they are initializing formals, declaring
+parameters, or even constructor parameters at all. This is because the language
+already allows using private names in all positional parameters, and we want a
+consistent user experience for all of them.*
```

Comment:
Maybe it's a good idea. Maybe it's inconsistent that it doesn't work for other kinds of named parameters too. I could see myself wanting to write `bar({int _foo = 0}) { ... }` to not shadow a surrounding `foo`, while still getting `foo` as parameter name.

I think it's an easier story to tell if it only works for initializing formals and field parameters, parameters whose name *also* refers to something else, and that other thing might want to be private. For everything else, if you choose a private name, it's because you wanted one, and you could just use a public name.